### DISCUSSION TITLE
With guix `emacs-magit` package: Required package ‘git-commit-20230101’ is unavailable

### DISCUSSION THREAD

**Discussion by ZelphirKaltstahl** [magit/magit#5208] [Support] [answered]
# Disclaimer

This is about  the Guix package for magit.

I am not sure this is the correct place to create an issue for this. If you are not usually updating the `emacs-magit` package on Guix, I will probably have to copy this whole issue into a message to the mailing list.

# Bug description

When installing Emacs via GNU Guix, one can also install Emacs packages via Guix. Also one can install `emacs-magit` in its newest version `4.0.0`. So far so good. However, when one runs an Emacs in a Guix profile with that `emacs-magit` installed and loads Guix Emacs packages in ones `init.el`, using:

```
;; Load packages, which are installed via guix.

(setq guix-package-enable-at-startup t)
(require 'guix-init nil t)
```

that installed `emacs-magit` package seems to be broken, and I think I have found the reason why that is (more later).

## (1) what behavior you expected

I expected the guix installed `emacs-magit` package to work flawlessly.

## (2) what behavior you observed

Emacs cannot load `magit` installed that way, because it runs into an error:

```
Required package ‘git-commit-20230101’ is unavailable
```

This is surprising for at least 2 reasons:

1. Guix should have installed all dependencies or failed, when the dependency does not exist.
2. Installing `magit` using `(use-package magit :ensure t :demand t)` does not run into this problem.
3. Installing `magit` using `(use-package magit :ensure t :demand t)` also depends on a newer `git-commit` version.

## (3) and how we can reproduce the issue.

I have a repo at <https://codeberg.org/ZelphirKaltstahl/learn-you-some-erlang-workthrough/src/commit/d0927d28a0ee57dfc139ed12af4980767611c85b>, where I try to make a directory local Emacs setup using Guix. In the scope of that repository I run into the problem. The readme lists the steps:

`guix-env/manifest.scm`:

```scheme
(specifications->manifest
 '("emacs"
   "emacs-magit"))
```

`guix-env/channels.scm`

```scheme
(list (channel
        (name 'guix)
        (url "https://git.savannah.gnu.org/git/guix.git")
        (branch "master")
        (commit
          "3c58b759a51072aabd7eaaca680674a0c2b36c23")
        (introduction
          (make-channel-introduction
            "9edb3f66fd807b096b48283debdcddccfea34bad"
            (openpgp-fingerprint
              "BBB0 2DDF 2CEA F6A8 0D1D  E643 A2A0 6DF2 A33A 54FA")))))
```

`emacs/init.el`:

```elisp
(setq guix-package-enable-at-startup t)
(require 'guix-init nil t)

(require 'package)
(add-to-list 'package-archives '("melpa-stable" . "https://stable.melpa.org/packages/") t)
(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
(package-initialize)

(use-package magit
  :ensure t
  :demand t)
```

Then running the following commands:

```shell
guix time-machine --channels=guix-env/channels.scm -- package --manifest=guix-env/manifest.scm --profile=guix-profile
GUIX_PROFILE="guix-profile" ; . "${GUIX_PROFILE}/etc/profile"
emacs --init-directory emacs --maximized --debug-init --no-site-lisp --no-splash --no-desktop --chdir "."
```

Will result in Emacs starting and showing:

```
Unable to activate package ‘magit’.
Required package ‘git-commit-20230101’ is unavailable
```

> Please include a backtrace in your report.  In most cases doing:

>     M-x toggle-debug-on-error RET

> and then going through the steps again should result in a backtrace.

This did not show me anything more in the messages buffer.

> Also post the output of:

>     M-x magit-version RET

Since magit could not be activated, this function is not available.

> Before reporting a defect please try to reproduce it using an Emacs instance in which only Magit and its dependencies have been loaded. Other packages or your configuration should not be loaded. This makes it easier to determine whether the issue lays with Magit or something else.

This is what my linked repository does, using GNU Guix.

> Finally, if that didn't work and you have installed Magit from Melpa, then run commands similar to the ones above, but use tab completion to replace the various Ns with the correct versions:
> 
>     $ cd ~/.emacs.d/elpa/magit-N
>     $ emacs -Q --debug-init --eval '(setq debug-on-error t)' -L ../dash-N -L ../git-commit-N -L ../transient-N -L ../with-editor-N -L ../ghub-N -L . -l magit

When I install magit from Melpa, I can clearly see Emacs logging compile messages about the dependencies listed here. However, if we check what is specified for the `emacs-magit` package on Guix, we see the following:

```shell
name: emacs-magit
version: 4.0.0
outputs:
+ out: everything
systems: x86_64-linux i686-linux
dependencies: emacs-compat@30.0.0.0 emacs-dash@2.19.1 emacs-transient@0.7.4 emacs-with-editor@3.4.1 git@2.45.2 perl@5.36.0 texinfo@6.8
location: gnu/packages/emacs-xyz.scm:1710:2
homepage: https://magit.vc/
license: GPL 3+
synopsis: Emacs interface for the Git version control system  
description: With Magit, you can inspect and modify your Git repositories with Emacs.  You can review and commit the changes you have made to the tracked files, for example, and you can browse the history of
+ past changes.  There is support for cherry picking, reverting, merging, rebasing, and other common Git operations.
relevance: 20
``` 

There is no `emacs-git-commit` dependency! It simply seems to be missing.

# Conclusion

When Emacs then tries to load a magit installed via Guix, it is missing `git-commit` and somehow tries to find an old version of if: `git-commit-20230101`, which it does not find.

# Fix

However, it is also not easily fixed by adding the dependency `emacs-git-commit` to `manifest.scm`, because no `emacs-git-commit` exists on Guix yet. I am guessing one would first have to create that package and then add it as dependency to `emacs-magit` to make the Guix package for magit work properly.

# Workaround

Not installing magit from guix but from Melpa helps. To do that one can simply:

* comment out `emacs-magit` in `manifest.scm`
* remove the guix profile and all files inside `emacs/`, except for `init.el`
* recreate the profile (this time it is without `emacs-magit`!)
* run emacs

It will install `magit` just fine.

**Comment by tarsius** [comment]
> ```
> (add-to-list 'package-archives '("melpa-stable" . "https://stable.melpa.org/packages/") t)
> (add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
> ```

You should not use these two channels together.

I strongly recommend you use `melpa` not `melpa-stable`, because "stable" is a misnomer, for many packages "stale" would be a better term. All Melpa maintainers recommend that users stick to `melpa`.

To ensure a clean slate, not only remove the entry for `melpa`, but also remove the respective directory from `~/.config/emacs/elpa`, restart emacs, and `M-x package-refresh-contents`. After that you probably won't get this error anymore.

**Reply by ZelphirKaltstahl** [reply]
I don't think this is the core of the problem with the guix package `emacs-magit`. I don't see how melpa things and `M-x package-refresh-contents` would be related to the guix package, because the guix package gets installed in the guix profile, which is created as described in the readme file.

It rather seems to me a missing dependency in the definition of the `emacs-magit` Guix package is the problem. Ideally I wouldn't even need any Melpa or Melpa stable package registries, since I could install magit via Guix. 

Partially I wanted to create awareness, that the guix package definition of `emacs-magit` might be incorrect/incomplete. Perhaps you know more about this package or are even the author?

Thanks for the advice about melpa stable though. I was not aware of this and will improve my Emacs configurations in general.

Nevertheless, I will try your suggestion and report.

**Reply by tarsius** [reply]
I think the core of the problem is that you have multiple versions of these packages installed and that you are not certain which is being used.

> ```
> Unable to activate package ‘magit’.
> Required package ‘git-commit-20230101’ is unavailable
> ```

These sound like errors from `package.el`, not guix.

While I do use guix, I don't currently use it for guix packages. But I've quickly installed magit (4.0.0) and looked at the installed files. Magit itself contains `git-commit.el`. That file is maintained in magit's repository, and while melpa currently splits that out into a separate package, guix does not. I will probably change that on melpa as well.

Deleting all of `~/.config/emacs/elpa` would probably fix this. To be on the save side don't delete it, just move it out of the way.

**Reply by ZelphirKaltstahl** [reply]
I will check this again and then write the results here.

> These sound like errors from package.el, not guix.

Yes, I am guessing this is, because packages installed via guix still need to be loaded, and at that time magit realizes, that there is no `git-commit` installed, so it fails. That's my theory at least.

Pretty sure I cleared the whole emacs dir (where the `init.el` and `elpa` and so on is contained in), and when I had `emacs-magit` in my `manifest.scm` I removed the `:ensure t` from my `use-package` call, to avoid installing it from melpa and vice versa, when I installed it from melpa, I commented it out in my `manifest.scm`.

I still find it curious, that the `git-commit` dependency is completely missing from the `dependencies` of the `emacs-magit` package on guix, though.

**Reply by tarsius** [reply]
> I still find it curious, that the `git-commit` dependency is completely missing from the `dependencies` of the `emacs-magit` package on guix, though.

That isn't curious at all. On guix `git-commit.el` is part of `emacs-magit`. Files that are part of a package do not have to be specified as dependencies of that package.

**Reply by ZelphirKaltstahl** [reply]
I tried again:

1. removed guix profile
2. commented in `emacs-magit` in `manifest.scm`
3. commented out `:ensure t` in `init.el`
4. recreated guix profile
5. activated guix profile
6. ran emacs

```
⛔ Error (use-package): Cannot load magit
```

In `*Warnings*` buffer and 

```
Unable to activate package ‘magit’.
Required package ‘git-commit-20230101’ is unavailable
```

In `*Messages*` buffer.

Then I:

1. closed emacs
2. moved `~/.config/emacs/elpa` to `~/.config/emacs/elpa.backup`
3. ran emacs again

Same error.
That makes sense, since I am specifying `--init-directory` when starting Emacs with activated guix profile and my init directory is in the project directory.

```shell
emacs \
    --init-directory emacs \
    --maximized \
    --debug-init \
    --no-site-lisp \
    --no-splash \
    --no-desktop \
    --chdir "."
```

So Emacs shouldn't even look in `~/.config/emacs/elpa`.

**Reply by tarsius** [reply, marked as answer]
I guess there is a problem with the guix package then. I've never used that myself, so I cannot help you with that. As you said, the guix mailing list might be the next place to ask for help.

**Reply by ZelphirKaltstahl** [reply]
OK, thanks nevertheless! I will take it there. Perhaps, if I can find a solution there, I can also post it here, in case anyone else is searching.
